### DISCUSSION TITLE
Add dunder method to modify class state before and after parsing/validation

### DISCUSSION THREAD

**Discussion by GlenNicholls** [pydantic/pydantic#4431] [Feature Request]
I've been dealing with custom types and it'd be really useful to have a post parse/validate method to modify internal state similar to the dataclass `__post_init__` and `__post_init_post_parse__`. The below is just a simple example:

``` python
from typing import Optional, List

def init(arg: str):
    return arg.split(".")


class A(str):
    _foo: Optional[List[str]] = None
 
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        # do stuff
        return cls(v)

    @property
    def foo(self):
        if self._foo is None:
            # Set foo based on value of self if it hasn't already been set 
            self._foo = init(self)
        return self._foo
```

I could add `__new__` to get around this, but I'm implementing a Pydantic compatible type and `__new__` doesn't help if I need the data to be parsed already:

``` python
class A(str):
    def __new__(arg, **kwargs):
        # I could also add custom args above, but for this case I don't need that so this seems like an ugly hack to get what I want
        super().__new__(arg, **kwargs)
        self._foo = init(arg)
...
```

I think the below would work although I'm not sure if `init` would set `_foo` on an instance of the class or the class itself. The docs lead me to believe it would be set on the class, but some quick tests seemed to indicate this worked okay:

``` python
from pydantic import BaseConfig, BaseModel, validator

class A(BaseModel):
    __root__: str

    _foo: str
    class Config:
        underscore_attrs_are_private = True

    @validator("__root__")
    def validate(cls, v):
        # do stuff
        return v

    @root_validator(skip_on_failure=True)
    def init(cls, values)
        # do stuff based on values["__root__"]
        cls._foo = X
        return values

    @property 
    def foo(self)
        return self._foo
```

Even if the above works as I expect, I'm using pydantic to parse config files and when the error message includes `__root__`, it is confusing for users because that is pointing to the model implementation and not the data in their config file directly.

So, I would love if Pydantic had some dunders to support this directly. Ideally, I would like to have `__post_parse__` that would work for both BaseModel and custom data types. I dislike the above because I end up cluttering code with comments explaining why it's so weird and very specific details about how to actually use the classes. This makes the code somewhat ugly, difficult to document, and provides limited options to choose the right tool. If a dataclass makes sense, that's perfect because everything I need exists. However, if I'm looking for a custom root type or custom data type, it just isn't easy to implement cleanly.

**Comment by samuelcolvin** [comment]
why not just yield another validator?

**Reply by GlenNicholls** [reply]
Hmm, can you be more specific about what you mean?

**Reply by GlenNicholls** [reply]
For reference, I get it to work when I modify an instance of the class:

``` python
    class FooBar(str):
        _double = None

        @classmethod
        def __get_validators__(cls):
            yield cls.validate

        @classmethod
        def validate(cls, v):
            # do stuff
            if v not in ["foo", "bar"]:
                raise ValueError("Value is not foo or bar")
            cls = cls(v)
            cls._double = v * 2
            return cls

    class A(BaseModel):
        n: FooBar

    a = A(n="foo")
    b = A(n="bar")
    print(a, a.n._double, b, b.n._double)
# n='foo' foofoo n='bar' barbar
```

But this doesn't feel right. I suppose the above will work, but that's kind of awkward. If the class is an ABC, the intent of the code would be much clearer if I could just do:

``` python
def __post_parse__(self):
    super().__post_parse__()
    self._foo = blah
    ....
```
